File: src/providers/openrouter.ts
import type { DecisionRequest } from "../types.js";
import { type FetchLike, HttpDecisionProvider } from "./http.js";

export interface OpenRouterProviderOptions {
  apiKey?: string;
  model?: string;
  endpoint?: string;
  appName?: string;
  appUrl?: string;
  fetch?: FetchLike;
  maxRetries?: number;
  timeoutMs?: number;
  inputCostPerMillion?: number;
  outputCostPerMillion?: number;
}

function normalizeModel(model: string): string {
  if (model === "jev-latest") return "typesafe/jev-1.13";
  if (model.startsWith("~typesafe/") || model.startsWith("typesafe/")) return model;
  return `typesafe/${model}`;
}

export class OpenRouterProvider extends HttpDecisionProvider {
  constructor(options: OpenRouterProviderOptions = {}) {
    const appName = options.appName ?? "DecideKit";
    const appUrl = options.appUrl ?? "https://www.npmjs.com/package/decidekit";
    const apiKey = options.apiKey ?? process.env.OPENROUTER_API_KEY ?? "";
    const fetchOption = options.fetch ? { fetch: options.fetch } : {};
    const retriesOption = options.maxRetries === undefined ? {} : { maxRetries: options.maxRetries };
    const timeoutOption = options.timeoutMs === undefined ? {} : { timeoutMs: options.timeoutMs };
    super({
      name: "openrouter",
      endpoint: options.endpoint ?? "https://openrouter.ai/api/alpha/decisions",
      apiKey,
      defaultModel: normalizeModel(options.model ?? "typesafe/jev-1.13"),
      headers: {
        "HTTP-Referer": appUrl,
        "X-OpenRouter-Title": appName,
      },
      ...fetchOption,
      ...retriesOption,
      ...timeoutOption,
      pricing: {
        inputCostPerMillion: options.inputCostPerMillion ?? 0.042,
        outputCostPerMillion: options.outputCostPerMillion ?? 0,
      },
      transformBody: (request: DecisionRequest, model: string) => ({
        model: normalizeModel(model),
        state: request.state,
        questions: request.questions,
      }),
    });
  }
}

File: python/decidekit/providers/openrouter.py
from __future__ import annotations

import os
from typing import Any

from ..types import DecisionQuestion, JsonValue
from .http import HttpDecisionProvider, Transport


def normalize_model(model: str) -> str:
    if model == "jev-latest":
        return "typesafe/jev-1.13"
    if model.startswith(("typesafe/", "~typesafe/")):
        return model
    return f"typesafe/{model}"


class OpenRouterProvider(HttpDecisionProvider):
    def __init__(
        self,
        api_key: str | None = None,
        model: str = "typesafe/jev-1.13",
        *,
        endpoint: str = "https://openrouter.ai/api/alpha/decisions",
        app_name: str = "DecideKit",
        app_url: str = "https://www.npmjs.com/package/decidekit",
        max_retries: int = 2,
        timeout: float = 30.0,
        input_cost_per_million: float = 0.042,
        output_cost_per_million: float = 0.0,
        transport: Transport | None = None,
    ):
        super().__init__(
            name="openrouter",
            endpoint=endpoint,
            api_key=api_key or os.environ.get("OPENROUTER_API_KEY", ""),
            default_model=normalize_model(model),
            headers={"HTTP-Referer": app_url, "X-OpenRouter-Title": app_name},
            max_retries=max_retries,
            timeout=timeout,
            input_cost_per_million=input_cost_per_million,
            output_cost_per_million=output_cost_per_million,
            transport=transport,
            transform_body=lambda state, questions, effective_model: {
                "model": normalize_model(effective_model),
                "state": state,
                "questions": {key: value.to_payload() for key, value in questions.items()},
            },
        )
